internal/provider/resource_offer.go

```diff
@@ -2,173 +2,255 @@ package provider
+	resp.Schema = frameworkResSchema.Schema{
+		Description: "A resource that represent a Juju Offer.",
+		Attributes: map[string]frameworkResSchema.Attribute{
+			"model": frameworkResSchema.StringAttribute{
 				Description: "The name of the model to operate in.",
-				Type:        schema.TypeString,
 				Required:    true,
-				ForceNew:    true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.RequiresReplace(),
+				},
 			},
-			"name": {
+			"name": frameworkResSchema.StringAttribute{
 				Description: "The name of the offer.",
-				Type:        schema.TypeString,
 				Optional:    true,
 				Computed:    true,
-				ForceNew:    true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.RequiresReplace(),
+				},
 			},
-			"application_name": {
+			"application_name": frameworkResSchema.StringAttribute{
 				Description: "The name of the application.",
-				Type:        schema.TypeString,
 				Required:    true,
-				ForceNew:    true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.RequiresReplace(),
+				},
 			},
-			"endpoint": {
+			"endpoint": frameworkResSchema.StringAttribute{
 				Description: "The endpoint name.",
-				Type:        schema.TypeString,
 				Required:    true,
-				ForceNew:    true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.RequiresReplace(),
+				},
 			},
-			"url": {
+			"url": frameworkResSchema.StringAttribute{
 				Description: "The offer URL.",
-				Type:        schema.TypeString,
 				Computed:    true,
 			},
+			"id": frameworkResSchema.StringAttribute{
+				Computed: true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.UseStateForUnknown(),
+				},
+			},
 		},
 	}
 }
 
-func resourceOfferCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
-	client := meta.(*juju.Client)
+func (o *offerResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	// Prevent panic if the provider has not been configured.
+	if o.client == nil {
+		addClientNotConfiguredError(&resp.Diagnostics, "offer", "create")
+		return
+	}
 
-	var diags diag.Diagnostics
-	modelName := d.Get("model").(string)
-	modelInfo, err := client.Models.GetModelByName(modelName)
-	if err != nil {
-		return diag.FromErr(err)
+	var data offerResourceModel
+
+	// Read Terraform configuration from the request into the model
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
 	}
 
+	modelName := data.ModelName.ValueString()
+	modelInfo, err := o.client.Models.GetModelByName(data.ModelName.ValueString())
+	if err != nil {
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to resolve model UUID, got error: %s", err))
+		return
+	}
 	modelUUID := modelInfo.UUID
 	modelOwner := strings.TrimPrefix(modelInfo.OwnerTag, juju.PrefixUser)
```

Comment:
Let's have a todo on this one. We're leaking juju-ism's out of its package in 2 ways, returning the params struct and not trimming the tag first. 

Tag.Id() is usually a better way to do this, rather than the trim.